spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -85,48 +109,148 @@ public Object postProcessAfterInitialization(final Object bean, String beanName)
-					}
-					if (mappedBindings.containsKey(streamListener.value())) {
-						throw new BeanInitializationException("Duplicate @" + StreamListener.class.getSimpleName() +
-								" mapping for '" + streamListener.value() + "' on " + invocableHandlerMethod.getShortLogMessage() +
-								" already existing for " + mappedBindings.get(streamListener.value()).getShortLogMessage());
+					Class<?>[] parameterTypes = method.getParameterTypes();
+					boolean hasInputOrOutputAnnotatedParameters = false;
+					for (int parameterIndex = 0; parameterIndex < parameterTypes.length; parameterIndex++) {
+						MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method,
+								parameterIndex);
+						boolean currentParameterIsAnnotated = methodParameter.getParameterAnnotation(
+								Input.class) != null || methodParameter.getParameterAnnotation(Output.class) != null;
+						if (hasInputOrOutputAnnotatedParameters) {
+							Assert.isTrue(currentParameterIsAnnotated,
+									"Either all parameters must be annotated with @Input and @Output, or none");
+						}
+						else {
+							hasInputOrOutputAnnotatedParameters = currentParameterIsAnnotated;
+						}
 					}
-					mappedBindings.put(streamListener.value(), invocableHandlerMethod);
-					SubscribableChannel channel = applicationContext.getBean(streamListener.value(),
-							SubscribableChannel.class);
-					final String defaultOutputChannel = extractDefaultOutput(method);
-					if (invocableHandlerMethod.isVoid()) {
-						Assert.isTrue(StringUtils.isEmpty(defaultOutputChannel), "An output channel cannot be specified for a method that " +
-								"does not return a value");
+					if (!hasInputOrOutputAnnotatedParameters) {
+						registerHandlerMethodOnListenedChannel(method, streamListener, bean);
 					}
 					else {
-						Assert.isTrue(!StringUtils.isEmpty(defaultOutputChannel), "An output channel must be specified for a method that " +
-								"can return a value");
+						invokeSetupMethodOnListenedChannel(method, bean);
 					}
-					StreamListenerMessageHandler handler = new StreamListenerMessageHandler(invocableHandlerMethod);
-					handler.setApplicationContext(applicationContext);
-					handler.setChannelResolver(binderAwareChannelResolver);
-					if (!StringUtils.isEmpty(defaultOutputChannel)) {
-						handler.setOutputChannelName(defaultOutputChannel);
-					}
-					handler.afterPropertiesSet();
-					channel.subscribe(handler);
 				}
 			}
 		});
 		return bean;
 	}
 
+	private void invokeSetupMethodOnListenedChannel(Method method, Object bean) {
+		Object[] arguments = new Object[method.getParameterTypes().length];
+		for (int parameterIndex = 0; parameterIndex < arguments.length; parameterIndex++) {
+			MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+			Class<?> parameterType = methodParameter.getParameterType();
+			Input input = methodParameter.getParameterAnnotation(Input.class);
+			if (input != null) {
```

Comment:
I'm stumped as to why the if and else branch appear identical, since the argument adapters themselves must be directional